src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -950,13 +973,36 @@ public SolutionCompilationState WithAnalyzerConfigDocumentText(
     /// <inheritdoc cref="Solution.WithDocumentSyntaxRoots(ImmutableArray{ValueTuple{DocumentId, SyntaxNode}}, PreservationMode)"/>
     public SolutionCompilationState WithDocumentSyntaxRoots(ImmutableArray<(DocumentId documentId, SyntaxNode root)> syntaxRoots, PreservationMode mode)
     {
-        return UpdateDocumentsInMultipleProjects<DocumentState, SyntaxNode, PreservationMode>(
-            syntaxRoots,
+        using var _1 = ArrayBuilder<(DocumentId, SyntaxNode)>.GetInstance(out var ordinaryDocuments);
+        using var _2 = ArrayBuilder<(SourceGeneratedDocumentIdentity documentIdentity, DateTime generationDateTime, SourceText text, SyntaxNode? syntaxNode)>.GetInstance(out var sourceGeneratedUpdates);
+        foreach (var doc in syntaxRoots)
+        {
+            if (!doc.documentId.IsSourceGenerated)
+            {
+                ordinaryDocuments.Add(doc);
+            }
+            else if (TryGetSourceGeneratedDocumentStateForAlreadyGeneratedId(doc.documentId) is { Identity: var identity })
+            {
+                // Source generated documents always do a full parse, and source generator authors are only allowed to provide
+                // strings, so we follow suit here and just take the text from the root.
+                sourceGeneratedUpdates.Add((identity, DateTime.UtcNow, doc.root.GetText(), doc.root));
```

Comment:
interesting... probably ok.  might blow up later.  not sure :D.

Reply:
IMO passing around SourceGeneratedStates would make the other callers of `WithFrozenSourceGeneratedDocuments` have to work a lot harder, and would end up having to do most of the work of that method here. Making source text optional could work, though it would just move this `doc.root.GetText()` into the `WithFrozenSourceGeneratedDocuments`, which isn't much of a difference IMO.

This was as nice as I could get it, but I did dream of a future where we had unions, because if we could do `(SourceText or SyntaxNode)`, I could have unified a bunch of this code.